src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -0,0 +1,254 @@
+{
+    UCollationResult result = UCOL_EQUAL;
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+
+    if (U_SUCCESS(err))
+    {
+        result = ucol_strcoll(pColl, lpStr1, cwStr1Length, lpStr2, cwStr2Length);
+        ucol_close(pColl);
+    }
+
+    return result;
+}
+
+/*
+Function:
+IndexOf
+*/
+extern "C" int32_t IndexOf(const char* lpLocaleName, const UChar* lpTarget, const UChar* lpSource, int32_t cwSourceLength, int32_t options)
+{
+    int32_t result = USEARCH_DONE;
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+
+    if (U_SUCCESS(err))
+    {
+        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, -1, lpSource, cwSourceLength, pColl, nullptr, &err);
+
+        if (U_SUCCESS(err))
+        {
+            result = usearch_first(pSearch, &err);
+            usearch_close(pSearch);
+        }
+
+        ucol_close(pColl);
+    }
+
+    return result;
+}
+
+/*
+Function:
+LastIndexOf
+*/
+extern "C" int32_t LastIndexOf(const char* lpLocaleName, const UChar* lpTarget, const UChar* lpSource, int32_t cwSourceLength, int32_t options)
+{
+    int32_t result = USEARCH_DONE;
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+
+    if (U_SUCCESS(err))
+    {
+        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, -1, lpSource, cwSourceLength, pColl, nullptr, &err);
+
+        if (U_SUCCESS(err))
+        {
+            result = usearch_last(pSearch, &err);
+            usearch_close(pSearch);
+        }
+
+        ucol_close(pColl);
+    }
+
+    return result;
+}
+
+/*
+ Return value is a "Win32 BOOL" (1 = true, 0 = false)
+ */
+extern "C" int32_t StartsWith(const char* lpLocaleName, const UChar* lpTarget, const UChar* lpSource, int32_t cwSourceLength, int32_t options)
+{
+    int32_t result = FALSE;
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+
+    if (U_SUCCESS(err))
+    {
+        UStringSearch* pSearch = usearch_openFromCollator(lpTarget, -1, lpSource, cwSourceLength, pColl, nullptr, &err);
+        int32_t idx = USEARCH_DONE;
+
+        if (U_SUCCESS(err))
+        {
+            idx = usearch_first(pSearch, &err);
+
+            if (idx == 0)
+            {
+                result = TRUE;
+            }
+            else
+            {
+                UCollationElements* pCollElem = ucol_openElements(pColl, lpSource, idx, &err);
+
+                if (U_SUCCESS(err))
+                {
+                    int32_t curCollElem = UCOL_NULLORDER;
+
+                    result = TRUE;
+
+                    while((curCollElem = ucol_next(pCollElem, &err)) != UCOL_NULLORDER)
```

Comment:
Ahh, I see what you are saying.  I will change it so that we don't check for U_FAILURE until after we finish the loop and return FALSE in that case as well.